treeherder/perf/alerts.py

```diff
@@ -334,52 +337,96 @@ def _glance_over_retrigger_range(self, data_points_to_retrigger: List[dict]):
         summaries_to_retrigger = self._fetch_by(
-            self._summaries(self.since),
+            self._summaries_requiring_reports(self.since),
             frameworks,
             repositories
         )
-        return self._identify_retriggerables(summaries_to_retrigger)
+        return self.compile_reports_for(summaries_to_retrigger)
+
+    def compile_reports_for(self, summaries_to_retrigger: QuerySet) -> List[BackfillReport]:
+        reports = []
+
+        for summary in summaries_to_retrigger:
+            important_alerts = self._pick_important_alerts(summary)
+            if len(important_alerts) == 0 and self._doesnt_have_report(summary):
+                continue  # won't create blank reports
+            # but will update if case
+
+            try:
+                alert_context_map = self._associate_retrigger_context(important_alerts)
+            except MissingRecords as ex:
+                self.log.warning(f"Failed to compute report for alert summary {summary}. {ex}")
+                continue
+
+            backfill_report, created = BackfillReport.objects.get_or_create(summary_id=summary.id)
+            if created or backfill_report.is_outdated:
+                self._provide_records(backfill_report, alert_context_map)
+            reports.append(backfill_report)
+
+        return reports
+
+    def _pick_important_alerts(self, from_summary: PerformanceAlertSummary) -> List[PerformanceAlert]:
+        return self.alerts_picker.extract_important_alerts(from_summary.alerts.all())
+
+    def _provide_records(self, backfill_report: BackfillReport,
+                         alert_context_map: List[Tuple]):
+        # any existing records are outdated; remove them
+        backfill_report.expel_records()
+
+        for alert, retrigger_context in alert_context_map:
+            BackfillRecord.objects.create(alert=alert,
+                                          report=backfill_report,
+                                          context=json.dumps(retrigger_context))
+
+    def _summaries_requiring_reports(self, timestamp: datetime) -> QuerySet:
+        recent_summaries_with_no_reports = Q(last_updated__gte=timestamp,
+                                             backfill_report__isnull=True)
+        summaries_with_outdated_reports = Q(last_updated__gt=F('backfill_report__last_updated'))
 
-    def _summaries(self, timestamp: datetime) -> QuerySet:
         return (PerformanceAlertSummary.objects
+                .prefetch_related('backfill_report')
                 .select_related('framework', 'repository')
-                .filter(created__gte=timestamp))
+                .filter(recent_summaries_with_no_reports | summaries_with_outdated_reports))
 
     def _fetch_by(self, summaries_to_retrigger: QuerySet, frameworks: List[str], repositories: List[str]) -> QuerySet:
         if frameworks:
             summaries_to_retrigger = summaries_to_retrigger.filter(framework__name__in=frameworks)
```

Comment:
Please make a better error handling. If frameworks is not provided (and can be, since is not checked before this call) the runtime exception would be something like: `NoneType has no attribute filter....`
Note: Mea culpa, I introduced this problem.

Reply:
No problem. Thanks for spotting this!